sqlmesh/core/dialect.py

```diff
@@ -755,12 +755,23 @@ def select_from_values_for_batch_range(
         exp.alias_(exp.cast(column, to=kind), column, copy=False)
         for column, kind in columns_to_types.items()
     ]
-    values_exp = exp.values(
-        [tuple(transform_values(v, columns_to_types)) for v in values[batch_start:batch_end]],
-        alias=alias,
-        columns=columns_to_types,
-    )
-    return exp.select(*casted_columns).from_(values_exp, copy=False)
+    select = exp.select(*casted_columns)
+
+    # Make sure we don't generate an empty VALUES clause
+    if values:
+        values_exp = exp.values(
+            [tuple(transform_values(v, columns_to_types)) for v in values[batch_start:batch_end]],
+            alias=alias,
+            columns=columns_to_types,
+        )
+        return select.from_(values_exp, copy=False)
+
+    # If there are no values, return a dummy query that always produces no rows
+    for projection in select.selects:
+        # There is no source to select from anymore so we swap out the cast columns with literals
+        projection.this.replace(exp.Literal.number(1))
+
+    return select.where(exp.false(), copy=False)
```

Comment:
let's just use copy false when things can get big, it litters the code base if we just put it everywhere

Reply:
sure

Reply:
Bigquery doesn't support a where clause without a from clause FYI

Reply:
Interesting, thanks @z3z1ma. Perhaps make a dummy single value `VALUES` clause? :P